File: python/peer.py
import bitarray
import connection
import message
import multiprocessing as mp
import os
import random
import socket
import repeating


def send_choke_and_unchoke(conn, to_choke, to_unchoke):
    if conn.other_peer_id in to_choke:
        conn.send(message.choke())
        conn.choked = True
    elif conn.other_peer_id in to_unchoke:
        conn.send(message.unchoke())
        conn.choked = False


class Peer(mp.Process):

    def __init__(self, peer_id, hostname, port, has_file, cfg, peer_info):
        super(Peer, self).__init__()
        self.peer_id = peer_id
        self.hostname = hostname
        self.port = port
        self.has_file = has_file
        self.peer_info = peer_info

        self.piece_size = cfg.get_int('PieceSize')
        self.file_size = cfg.get_int('FileSize')
        self.optim_unchoking_int = cfg.get_int('OptimisticUnchokingInterval')
        self.unchoking_int = cfg.get_int('UnchokingInterval')
        self.np = cfg.get_int('NumberOfPreferredNeighbors')

        self.num_pieces = self.file_size // self.piece_size + (0 if self.piece_size % self.file_size == 0 else 1)
        self.arr = bitarray.bitarray(self.num_pieces)
        self.pieces = {}
        self.connections = []
        self.optim_unchoked_id = -1

        if self.has_file:
            self.filename = cfg.get_str('FileName')
            with open(self.filename, 'rb') as f:
                for i in range(self.num_pieces):
                    self.pieces[i] = f.read(self.piece_size)
            self.arr.setall(True)
        else:
            folder = f'./peer_{self.peer_id}'
            if not os.path.exists(folder):
                os.mkdir(folder)
            self.filename = os.path.join(folder, cfg.get_str('FileName'))
            self.arr.setall(False)

    def run(self):
        repeating.Timer(self.optim_unchoking_int, self._optimistic_unchoke).start()
        repeating.Timer(self.unchoking_int, self._unchoke).start()

        for hostname, port in self.peer_info:
            sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            sock.connect((hostname, port))
            self.connections.append(connection.Connection(sock, True, self))
            self.connections[-1].start()

        with socket.socket() as server_sock:
            server_sock.bind((self.hostname, self.port))
            server_sock.listen()
            while True:
                sock, address = server_sock.accept()
                self.connections.append(connection.Connection(sock, False, self))
                self.connections[-1].start()

    def _optimistic_unchoke(self):
        choked_but_interested_ids = [conn.other_peer_id for conn in self.connections if conn.interested and conn.choked]
        if len(choked_but_interested_ids) > 0:
            rand_id = random.choice(choked_but_interested_ids)
            if rand_id != self.optim_unchoked_id:
                for conn in self.connections:
                    send_choke_and_unchoke(conn, [self.optim_unchoked_id], [rand_id])
            self.optim_unchoked_id = rand_id

    def _unchoke(self):
        unchoked_neighbor_ids = [conn.other_peer_id for conn in self.connections if not conn.choked]
        interested_neighbor_ids = [conn.other_peer_id for conn in self.connections if conn.interested]
        useful_conn = [conn for conn in self.connections if conn.pieces_received > 0]
        if self.has_file == 1 and len(unchoked_neighbor_ids) < self.np - 1 and len(interested_neighbor_ids) >= self.np - 1:
            random.shuffle(interested_neighbor_ids)
            to_choke = []
            to_unchoke = interested_neighbor_ids[0:self.np - 1]
        elif len(useful_conn) >= self.np - 1:
            top_connections = sorted(useful_conn, key=lambda conn: conn.pieces_received)
            top_ids = [conn.other_peer_id for conn in top_connections[:self.np - 1]]
            to_choke = list(set(unchoked_neighbor_ids) - set(top_ids))
            to_unchoke = list(set(top_ids) - set(unchoked_neighbor_ids))
        for conn in self.connections:
            conn.pieces_received = 0
            send_choke_and_unchoke(conn, to_choke, to_unchoke)

    def missing_pieces(self):
        return self.arr.count() < self.num_pieces

    def neighbor_missing_pieces(self):
        return any([conn.other_arr.count() < self.num_pieces for conn in self.connections])

File: java2/Peer.java
import java.io.FileInputStream;
import java.io.IOException;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.ArrayList;
import java.util.BitSet;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;

public class Peer implements Runnable {
    private final ExecutorService threadPool = Executors.newCachedThreadPool();
    private final ScheduledExecutorService scheduledThreadPool = Executors.newScheduledThreadPool(2);

    public final int peerId;
    public final String hostname;
    public final int port;
    public final int hasFile;
    public final String filename;
    public final int pieceSize;
    public final int numPieces;
    public final BitSet bitarray;
    public final Map<Integer, byte[]> pieces = new ConcurrentHashMap<>();
    public final List<Connection> connections = new CopyOnWriteArrayList<>();

    private final List<Peer> peers = new ArrayList<>();
    private final int numPrefNeighbors;
    private final int optimisticUnchokingInterval;
    private final int unchokingInterval;

    private int optimisticallyUnchokedId;

    public Peer(int peerId, String hostname, int port, int hasFile, Config config, List<Peer> peers) {
        this.peerId = peerId;
        this.hostname = hostname;
        this.port = port;
        this.hasFile = hasFile;
        this.peers.addAll(peers);

        numPrefNeighbors = config.getInt("NumberOfPreferredNeighbors");
        optimisticUnchokingInterval = config.getInt("OptimisticUnchokingInterval");
        unchokingInterval = config.getInt("UnchokingInterval");
        pieceSize = config.getInt("PieceSize");
        filename = config.getString("FileName");

        int fileSize = config.getInt("FileSize");
        numPieces = fileSize / pieceSize + (fileSize % pieceSize == 0? 0 : 1);
        bitarray = new BitSet(numPieces);
        bitarray.clear();

        if (hasFile == 1) {
            readFile();
            bitarray.set(0, numPieces, true);
        }
    }

    private void readFile() {
        try {
            final FileInputStream fileInputStream = new FileInputStream(filename);
            for (int i = 0; i < numPieces; i++) {
                byte[] piece = new byte[pieceSize];
                fileInputStream.read(piece);
                pieces.put(i, piece);
            }
        } catch (IOException e) {
            System.err.println(e.toString());
        }
    }

    public void run() {
        try {
            scheduledThreadPool.scheduleAtFixedRate(
                    this::optimisticUnchoke, 0, optimisticUnchokingInterval, TimeUnit.SECONDS);
            scheduledThreadPool.scheduleAtFixedRate(
                    this::unchoke, 0, unchokingInterval, TimeUnit.SECONDS);
            for (Peer peer : peers) {
                Connection conn = new Connection(true, new Socket(peer.hostname, peer.port), this);
                connections.add(conn);
                threadPool.execute(conn);
                System.out.printf("Peer %s connected to Peer %s%n", peerId, peer.peerId);
            }
            ServerSocket serverSocket = new ServerSocket(port);
            while (true) {
                Connection conn = new Connection(false, serverSocket.accept(), this);
                connections.add(conn);
                threadPool.execute(conn);
            }
        } catch (IOException e) {
            System.err.println(e.toString());
        }
    }

    private void optimisticUnchoke() {
        List<Integer> chokedButInterestedIds = connections.stream()
                .filter(conn -> conn.interested && conn.choked).map(conn -> conn.otherPeerId)
                .collect(Collectors.toList());
        if (!chokedButInterestedIds.isEmpty()) {
            int randInt = ThreadLocalRandom.current().nextInt(0, chokedButInterestedIds.size());
            int randId = chokedButInterestedIds.get(randInt);
            if (randId != optimisticallyUnchokedId) {
                for (Connection conn : connections) {
                    sendChokeAndUnchoke(conn, List.of(optimisticallyUnchokedId), List.of(randId));
                }
            }
            optimisticallyUnchokedId = randId;
        }
    }

    private void unchoke() {
        List<Integer> unchokedNeighborIds = connections.stream()
                .filter(conn -> !conn.choked).map(conn -> conn.otherPeerId)
                .collect(Collectors.toList());
        List<Integer> interestedNeighborIds = connections.stream()
                .filter(conn -> conn.interested).map(conn -> conn.otherPeerId)
                .collect(Collectors.toList());
        List<Connection> usefulConnections = connections.stream()
                .filter(conn -> conn.piecesReceived > 0)
                .collect(Collectors.toList());
        List<Integer> toChoke = new ArrayList<>();
        List<Integer> toUnchoke = new ArrayList<>();
        if (hasFile == 1 && unchokedNeighborIds.size() < numPrefNeighbors - 1
                && interestedNeighborIds.size() >= numPrefNeighbors - 1) {
            Collections.shuffle(interestedNeighborIds);
            toUnchoke.addAll(interestedNeighborIds.subList(0, numPrefNeighbors - 1));
        } else if (usefulConnections.size() >= numPrefNeighbors - 1) {
            connections.sort(Comparator.comparingInt(conn -> conn.piecesReceived));
            List<Integer> topIds = connections.subList(0, numPrefNeighbors - 1).stream()
                    .map(conn -> conn.otherPeerId).collect(Collectors.toList());
            toChoke.addAll(difference(unchokedNeighborIds, topIds));
            toUnchoke.addAll(difference(topIds, unchokedNeighborIds));
        }
        for (Connection conn : connections) {
            conn.piecesReceived = 0;
            sendChokeAndUnchoke(conn, toChoke, toUnchoke);
        }
    }

    private void sendChokeAndUnchoke(Connection conn, List<Integer> toChoke, List<Integer> toUnchoke) {
        try {
            if (toChoke.contains(conn.otherPeerId)) {
                conn.send(Message.choke());
                conn.choked = true;
            } else if (toUnchoke.contains(conn.otherPeerId)) {
                conn.send(Message.unchoke());
                conn.choked = false;
            }
        } catch (IOException e) {
            System.err.println(e);
        }
    }

    private List<Integer> difference(List<Integer> a, List<Integer> b) {
        List<Integer> diff = new ArrayList<>(a);
        diff.removeAll(b);
        return diff;
    }

    public boolean missingPieces() {
        synchronized (bitarray) {
            return bitarray.cardinality() < numPieces;
        }
    }

    public boolean neighborsMissingPieces() {
        for (Connection conn : connections) {
            if (conn.otherBitarray.cardinality() < numPieces) {
                return true;
            }
        }
        return false;
    }
}
